File: scripts/youngcha/distribution/inlinegearbox.js
const rotL = require("libraries/rotpowerlib");

const inlinegearbox = rotL.torqueExtend(Block, Building, "inline-gearbox",rotL.baseTypes.torqueConnector, {
	
	load(){
		this.super$load();
		this.topsprite = Core.atlas.find(this.name + "-top");
		this.overlaysprite = Core.atlas.find(this.name + "-overlay");
		this.moving = Core.atlas.find(this.name + "-moving");
		this.base = Core.atlas.find(this.name + "-base");
		this.mbase = Core.atlas.find(this.name + "-mbase");
		this.gear = Core.atlas.find(this.name + "-gear");
		
	},
	
},{
	
	updatePre()
	{
		this.getGraphConnector("torque graph").setInertia(20);
		this.getGraphConnector("torque graph").setFriction(0.02);
	},
	
	draw() {
		let torquegraph = this.getGraphConnector("torque graph");
		let shaftrot = torquegraph.getRotation();
		
		let fixedrot = ((this.rotdeg()+90)%180)-90;
		Draw.rect(inlinegearbox.base, this.x, this.y, 0);
		Draw.rect(inlinegearbox.mbase, this.x, this.y, fixedrot);
		//speeeeeeen
		let offset = rotL.dirs[(this.rotation+1)%4];
		let ox = offset.x*4;
		let oy = offset.y*4;
		rotL.drawRotRect(inlinegearbox.moving, this.x+ox, this.y+oy, 16, 14/4, 8, fixedrot, shaftrot, shaftrot+90);
		rotL.drawRotRect(inlinegearbox.moving, this.x-ox, this.y-oy, 16, 14/4, 8, fixedrot, shaftrot+90, shaftrot+180);
		//gears
		Draw.rect(inlinegearbox.gear, this.x+2, this.y+2, shaftrot);
		Draw.rect(inlinegearbox.gear, this.x-2, this.y+2, -shaftrot);
		Draw.rect(inlinegearbox.gear, this.x+2, this.y-2, -shaftrot);
		Draw.rect(inlinegearbox.gear, this.x-2, this.y-2, shaftrot);
		//
		Draw.rect(inlinegearbox.overlaysprite, this.x, this.y, fixedrot);
		Draw.rect(inlinegearbox.topsprite, this.x, this.y, fixedrot);
        this.drawTeamTop();
	}
	
	
});

inlinegearbox.rotate = true;
inlinegearbox.update = true;
inlinegearbox.solid = true;
inlinegearbox.getGraphConnectorBlock("torque graph").setAccept([1,1,0,0,1,1,0,0]);
inlinegearbox.getGraphConnectorBlock("torque graph").setBaseFriction(0.02);
inlinegearbox.getGraphConnectorBlock("torque graph").setBaseInertia(20);

File: main/src/unity/world/blocks/distribution/InlineGearbox.java
package unity.world.blocks.distribution;

import arc.graphics.g2d.*;
import arc.math.geom.*;
import unity.graphics.*;
import unity.world.blocks.*;

import static arc.Core.*;

public class InlineGearbox extends GraphBlock{
    TextureRegion topRegion, overlayRegion, movingRegion, baseRegion, mbaseRegion, gearRegion;//topsprite,overlaysprite,moving,base,mbase,gear

    public InlineGearbox(String name){
        super(name);
        rotate = solid = true;
    }

    @Override
    public void load(){
        super.load();
        topRegion = atlas.find(name + "-top");
        overlayRegion = atlas.find(name + "-overlay");
        movingRegion = atlas.find(name + "-moving");
        baseRegion = atlas.find(name + "-base");
        mbaseRegion = atlas.find(name + "-mbase");
        gearRegion = atlas.find(name + "-gear");
    }

    public class InlineGearboxBuild extends GraphBuild{
        @Override
        public void draw(){
            float shaftRot = torque().getRotation();
            float fixedRot = (rotdeg() + 90f) % 180f - 90f;
            Draw.rect(baseRegion, x, y);
            Draw.rect(mbaseRegion, x, y, fixedRot);
            Point2 offset = Geometry.d4(rotation + 1);
            float ox = offset.x * 4f;
            float oy = offset.y * 4f;
            UnityDrawf.drawRotRect(movingRegion, x + ox, y + oy, 16f, 3.5f, 8f, fixedRot, shaftRot, shaftRot + 90f);
            UnityDrawf.drawRotRect(movingRegion, x - ox, y - oy, 16f, 3.5f, 8f, fixedRot, shaftRot + 90f, shaftRot + 180f);

            Draw.rect(gearRegion, x + 2f, y + 2f, shaftRot);
            Draw.rect(gearRegion, x - 2f, y + 2f, -shaftRot);
            Draw.rect(gearRegion, x + 2f, y - 2f, -shaftRot);
            Draw.rect(gearRegion, x - 2f, y - 2f, shaftRot);

            Draw.rect(overlayRegion, x, y, fixedRot);
            Draw.rect(topRegion, x, y, fixedRot);
            drawTeamTop();
        }
    }
}
